When dispatching, send ActiveSnapshot along, not some random snapshot.

hlinnaka opened the pull request:
If the caller specifies DF_WITH_SNAPSHOT, so that the command is dispatched
to the segments with a snapshot, but it currently has no active snapshot in
the QD itself, that seems like a mistake.

In qdSerializeDtxContextInfo(), the comment talked about which snapshot to
use when the transaction has already been aborted. I didn't quite
understand that. I don't think the function is used to dispatch the "ABORT"
statement itself, and we shouldn't be dispatching anything else in an
already-aborted transaction.

This makes it more clear which snapshot is dispatched along with the command.
In theory, the latest or serializable snapshot can be different from the one
being used when the command is dispatched, although I'm not sure if there
are any such cases in practice.

In the upcoming 8.4 merge, there are more changes coming up to snapshot
management, which make it more difficult to get hold of the latest acquired
snapshot in the transaction, so changing this now will ease the pain of
merging that.

I'm not sure why, but after making the change in qdSerializeDtxContextInfo(),
I started to get a lot of "Too many distributed transactions for snapshot
(maxCount %d, count %d)" errors. Looking at the code, I don't understand
how it ever worked. I don't see any no guarantee that the array in
TempQDDtxContextInfo or TempDtxContextInfo was pre-allocated correctly.
Or maybe it got allocated big enough to hold max_prepared_xacts, which
was always large enough, but it seemed rather haphazard to me. So in
the spirit of "if you don't understand it, rewrite it until you do", I
changed the way the allocation of the inProgressXidArray array works.
In statically allocated snapshots, i.e. SerializableSnapshot and
LatestSnapshot, the array is malloc'd. In a snapshot copied with
CopySnapshot(), it is points to a part of the palloc'd space for the
snapshot. Nothing new so far, but I changed CopySnapshot() to set "maxCount"
to -1 to indicate that it's not malloc'd. Then I modified
DistributedSnapshot_Copy and DistributedSnapshot_Deserialize to not give up
if the target array is not large enough, but enlarge it as needed. Finally,
I made a little optimization in GetSnapshotData() when running in a QE, to
move the copying of the distributed snapshot data to outside the section
guarded by ProcArrayLock. ProcArrayLock can be heavily contended, so that's
a nice little optimization anyway, but especially now that
DistributedSnapshot_Copy() might need to realloc the array.

hlinnaka commented:
`gpcheckcat` is failing with this. The reason boils down to this:

If you run the following command, is there are row in the resulting temp table or not:

```
CREATE TEMPORARY TABLE mytemp AS SELECT oid FROM pg_class WHERE relname='mytemp';
```
In other words, does the SELECT see the pg_class entry for the new table that is being created in the same command? On master, without this patch, there is a row, i.e. it does see itself. After this patch, there isn't. I'm inclined to say that the new behavior is correct, because that's how PostgreSQL behaves as well. But it means that we'll need to fix gpcheckcat to not get confused by that.

ashwinstar commented:
On Tue, Oct 17, 2017 at 10:13 AM, Heikki Linnakangas <
notifications@github.com> wrote:

> gpcheckcat is failing with this. The reason boils down to this:
>
> If you run the following command, is there are row in the resulting temp
> table or not:
>
> CREATE TEMPORARY TABLE mytemp AS SELECT oid FROM pg_class WHERE relname='mytemp';
>
> In other words, does the SELECT see the pg_class entry for the new table
> that is being created in the same command? On master, without this patch,
> there is a row, i.e. it does see itself. After this patch, there isn't. I'm
> inclined to say that the new behavior is correct, because that's how
> PostgreSQL behaves as well. But it means that we'll need to fix gpcheckcat
> to not get confused by that.
>
I agree new behavior seems fine, more consistent and less confusing.

hlinnaka commented:
On 10/17/2017 08:40 PM, Ashwin Agrawal wrote:
> I agree new behavior seems fine, more consistent and less confusing.

Added a commit to rewrite the gpcheckcat query to not use a temp table.

- Heikki

ashwinstar commented:
@hlinnaka Please check historical, there were performance reasons why simpler queries where changed to use TEMP table and gpcheckcat needs to run as fast as possible to reduce downtimes in field.

ashwinstar commented on src/backend/cdb/cdbdistributedsnapshot.c:
```diff
@@ -358,26 +354,27 @@ DistributedSnapshot_Copy(
+	if (target->inProgressXidArray == NULL)
 	{
-		Assert(target->inProgressXidArray == NULL);
+		int			maxCount = source->count + 10;
```
+10 seems pretty random.

hlinnaka commented:
Ashwin dug into the historical GPDB repositroy and bug tracking system, and dug out some details:

The change to use a temporary table in this query was made, after a report that checking pg_attribute table took very long, on a cluster with over 1000 segments, where the total size of pg_attribute was over 200 GB. So we do need to worry about the performance of this.

I twiddled some more on the query, and arrived at this even simpler version:
```
postgres=# explain SELECT attrelid,attname,
                 case when master is null then segids
                      when segids is null then array[master.segid]
                      else master.segid || segids end as segids
          FROM
          (
            SELECT attrelid,attname, array_agg(gp_segment_id order by gp_segment_id) as segids
            FROM gp_dist_random('pg_attribute')  GROUP BY attrelid,attname
          ) as seg
          FULL OUTER JOIN
          (
            SELECT gp_segment_id as segid, attrelid,attname FROM pg_attribute 
          ) as master
          USING (attrelid,attname)
          WHERE master.segid is null
             OR segids is null
             OR NOT segids @> (select array_agg(content::int4) from gp_segment_configuration WHERE content >= 0)
;
                                                                                                                                         QUERY PLAN                                           
                                                                                               
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
-----------------------------------------------------------------------------------------------
 Gather Motion 3:1  (slice3; segments: 3)  (cost=1000025803.02..1000029768.05 rows=63147 width=204)
   ->  Merge Full Join  (cost=1000025803.02..1000029768.05 rows=21049 width=204)
         Merge Cond: pg_catalog.pg_attribute.attrelid = pg_catalog.pg_attribute.attrelid AND pg_catalog.pg_attribute.attname = pg_catalog.pg_attribute.attname
         Filter: pg_catalog.pg_attribute.gp_segment_id IS NULL OR (array_agg(pg_catalog.pg_attribute.gp_segment_id ORDER BY pg_catalog.pg_attribute.gp_segment_id)) IS NULL OR (NOT (array_agg
(pg_catalog.pg_attribute.gp_segment_id ORDER BY pg_catalog.pg_attribute.gp_segment_id)) @> $0)
         InitPlan 1 (returns $0)  (slice4)
           ->  Aggregate  (cost=1.00..1.02 rows=1 width=32)
                 ->  Seq Scan on gp_segment_configuration  (cost=0.00..1.00 rows=1 width=2)
                       Filter: content >= 0
         ->  GroupAggregate  (cost=19560.58..22006.04 rows=21054 width=100)
               Group By: pg_catalog.pg_attribute.attrelid, pg_catalog.pg_attribute.attname
               ->  Sort  (cost=19560.58..19974.56 rows=55198 width=72)
                     Sort Key: pg_catalog.pg_attribute.attrelid, pg_catalog.pg_attribute.attname
                     ->  Redistribute Motion 3:3  (slice1; segments: 3)  (cost=0.00..5205.82 rows=55198 width=72)
                           Hash Key: pg_catalog.pg_attribute.attrelid, pg_catalog.pg_attribute.attname
                           ->  Seq Scan on pg_attribute  (cost=0.00..1893.94 rows=55198 width=72)
         ->  Sort  (cost=6241.43..6379.42 rows=18400 width=104)
               Sort Key: pg_catalog.pg_attribute.attrelid, pg_catalog.pg_attribute.attname
               ->  Redistribute Motion 1:3  (slice2)  (cost=0.00..1893.94 rows=55198 width=104)
                     Hash Key: pg_catalog.pg_attribute.attrelid, pg_catalog.pg_attribute.attname
                     ->  Seq Scan on pg_attribute  (cost=0.00..789.98 rows=55198 width=104)
 Settings:  optimizer=off
 Optimizer status: legacy query optimizer
(22 rows)
```

This was on my laptop, and on a small cluster, but I think this is as good as that query is going to get. None of the operators need a lot of memory, so I'm pretty sure you'd get the same plan even on a large system. I believe this is not worse than the old query.

This produces a slightly different result than the old one, because I moved some of the post-processing out of the SQL query, into the Python code. See updated commit in this PR.

ashwinstar commented on src/backend/cdb/cdbdistributedsnapshot.c:
```diff
@@ -326,22 +326,18 @@ DistributedSnapshot_Reset(DistributedSnapshot *distributedSnapshot)
 	distributedSnapshot->count = 0;
 
 	/* maxCount and inProgressXidArray left untouched */
+	if (distributedSnapshot->maxCount < 0)
```
>> Nothing new so far, but I changed CopySnapshot() to set
"maxCount" to -1 to indicate that it's not malloc'd.

Don't really like this hack and over the period this would become pretty confusing as well. 

It seems no more need `maxCount`, just `count` is enough and instead have explicit boolean to really know malloc'ed/palloc'ed area.

hlinnaka replied:
Hmm. Without `maxCount`, how do you know if you need to enlarge the array or not?

ashwinstar commented on src/backend/storage/ipc/procarray.c:
```diff
@@ -1413,7 +1411,18 @@ GetSnapshotData(Snapshot snapshot, bool serializable)
-	FillInDistributedSnapshot(snapshot);
+	if (DistributedTransactionContext == DTX_CONTEXT_QD_DISTRIBUTED_CAPABLE)
+	{
+		snapshot->haveDistribSnapshot = createDtxSnapshot(&snapshot->distribSnapshotWithLocalMapping);
```
While at it, lets rename this to `CreateDistributedSnapshot()` to be little consistent.

ashwinstar commented on src/backend/storage/ipc/procarray.c:
```diff
@@ -1510,6 +1519,13 @@ GetSnapshotData(Snapshot snapshot, bool serializable)
 	LWLockRelease(ProcArrayLock);
 
 	/*
+	 * If we didn't need to fill in the snapshot while holding the lock, do it now.
```
Please rewrite this comment to be more clear.

ashwinstar commented:
On Wed, Oct 18, 2017 at 7:27 AM, Heikki Linnakangas <
notifications@github.com> wrote:

> *@hlinnaka* commented on this pull request.
> ------------------------------
>
> In src/backend/cdb/cdbdistributedsnapshot.c
> <https://github.com/greenplum-db/gpdb/pull/3539#discussion_r145431452>:
>
> > @@ -326,22 +326,18 @@ DistributedSnapshot_Reset(DistributedSnapshot *distributedSnapshot)
>  	distributedSnapshot->count = 0;
>
>  	/* maxCount and inProgressXidArray left untouched */
> +	if (distributedSnapshot->maxCount < 0)
>
> Hmm. Without maxCount, how do you know if you need to enlarge the array
> or not?
>
Diff between `count` and `maxcount` is just +10 right. So, doesn't seem
worth having it, `count` can just serve the purpose, right ?

hlinnaka commented:
On 10/18/2017 05:31 PM, Ashwin Agrawal wrote:
> On Wed, Oct 18, 2017 at 7:27 AM, Heikki Linnakangas <
> notifications@github.com> wrote:
> 
>> *@hlinnaka* commented on this pull request.
>> ------------------------------
>>
>> In src/backend/cdb/cdbdistributedsnapshot.c
>> <https://github.com/greenplum-db/gpdb/pull/3539#discussion_r145431452>:
>>
>>> @@ -326,22 +326,18 @@ DistributedSnapshot_Reset(DistributedSnapshot *distributedSnapshot)
>>   	distributedSnapshot->count = 0;
>>
>>   	/* maxCount and inProgressXidArray left untouched */
>> +	if (distributedSnapshot->maxCount < 0)
>>
>> Hmm. Without maxCount, how do you know if you need to enlarge the array
>> or not?
>>
> Diff between `count` and `maxcount` is just +10 right. So, doesn't seem
> worth having it, `count` can just serve the purpose, right ?

The number of items can fluctuate from one call to another. So 'count' 
can be 1 on first call, 10 on next one, then 1 again, and so on. If we 
don't keep track of the allocated size, then we're forced to call 
realloc whenever the new count is larger than on previous call.

Furthermore, DistributedSnapshot_Reset() actually resets 'count' to 0, 
so we won't even know the size of the last call anymore, so we'd have to 
always call realloc.

- Heikki

ashwinstar commented:
On Wed, Oct 18, 2017 at 7:35 AM, Heikki Linnakangas <
notifications@github.com> wrote:

> On 10/18/2017 05:31 PM, Ashwin Agrawal wrote:
> > On Wed, Oct 18, 2017 at 7:27 AM, Heikki Linnakangas <
> > notifications@github.com> wrote:
> >
> >> *@hlinnaka* commented on this pull request.
> >> ------------------------------
> >>
> >> In src/backend/cdb/cdbdistributedsnapshot.c
> >> <https://github.com/greenplum-db/gpdb/pull/3539#discussion_r145431452>:
> >>
> >>> @@ -326,22 +326,18 @@ DistributedSnapshot_Reset(DistributedSnapshot
> *distributedSnapshot)
> >> distributedSnapshot->count = 0;
> >>
> >> /* maxCount and inProgressXidArray left untouched */
> >> + if (distributedSnapshot->maxCount < 0)
> >>
> >> Hmm. Without maxCount, how do you know if you need to enlarge the array
> >> or not?
> >>
> > Diff between `count` and `maxcount` is just +10 right. So, doesn't seem
> > worth having it, `count` can just serve the purpose, right ?
>
> The number of items can fluctuate from one call to another. So 'count'
> can be 1 on first call, 10 on next one, then 1 again, and so on. If we
> don't keep track of the allocated size, then we're forced to call
> realloc whenever the new count is larger than on previous call.
>
> Furthermore, DistributedSnapshot_Reset() actually resets 'count' to 0,
> so we won't even know the size of the last call anymore, so we'd have to
> always call realloc.
>

Okay, seems fine then. +10 feels little random, lets atleast make it
#define so can be easily modified if required in future in one place. Also,
adding extra boolean to know malloc vs palloc would be better than
overloading on maxCount.

hlinnaka commented:
@ashwinstar, addressed your comments. I still didn't change the maxCount=-1 thing. I don't think a separate boolean would be an improvement. Then you'd just need to check the boolean instead of the maxCount=-1 special case everywhere.

ashwinstar commented:
@Chibin to review the gphceckcat change and also maybe point to tests to run validate its fine.

@hlinnaka I am still not sure on gpcheckcat query change how will it perform for large number of segments. Strongly feel since previously issue was seen in field its better to verify out its not causing regression. Was discussing with @Eulerizeit about this to confirm and get numbers to validate its fine change with large schemas with many segments.

Eulerizeit commented:
@pivotal-mike

asimrp commented on src/backend/cdb/dispatcher/cdbdisp_dtx.c:
```diff
@@ -185,89 +185,17 @@ qdSerializeDtxContextInfo(int *size, bool wantSnapshot, bool inCursor,
-			SerializableSnapshot == NULL && !IsAbortInProgress())
-		{
-			/*
-			 * unfortunately, the dtm issues a select for prepared xacts at
```
This comment was incorrect anyway.  DTM recovery invokes this function with `wantSnapshot = false`: `gatherRMInDoubtTransactions() --> CdbDispatchCommand(flags = FD_NONE)`.

It makes sense to dispatch select on `pg_prepared_xact()` UDF without a snapshot.  The UDF is not accessing any user table, simply returning contents of shared memory.

@hlinnaka I think the following `if` block should be removed from `PostgresMain()`:
```C
                    if (TempDtxContextInfo.distributedSnapshot.maxCount == 0)
			TempDtxContextInfo.distributedSnapshot.maxCount = max_prepared_xacts;

                    setupQEDtxContext(&TempDtxContextInfo);
```
Does that also explain the "Too many distributed transactions for snapshot" error you saw?

hlinnaka commented:
On 10/22/2017 11:09 PM, Asim RP wrote:
> asimrp commented on this pull request.
> 
> 
> 
>>   	 */
>   	*size = 0;
>   
>   	if (wantSnapshot)
>   	{
> -
> -		if (LatestSnapshot == NULL &&
> -			SerializableSnapshot == NULL && !IsAbortInProgress())
> -		{
> -			/*
> -			 * unfortunately, the dtm issues a select for prepared xacts at
> 
> This comment was incorrect anyway.  DTM recovery invokes this function with `wantSnapshot = false`: `gatherRMInDoubtTransactions() --> CdbDispatchCommand(flags = FD_NONE)`.

Ok, thanks for checking.

> It makes sense to dispatch select on `pg_prepared_xact()` UDF without a snapshot.  The UDF is not accessing any user table, simply returning contents of shared memory.
> 
> @hlinnaka I think the following `if` block should be removed from `PostgresMain()`:
> ```C
>                      if (TempDtxContextInfo.distributedSnapshot.maxCount == 0)
> 			TempDtxContextInfo.distributedSnapshot.maxCount = max_prepared_xacts;
> 
>                      setupQEDtxContext(&TempDtxContextInfo);
> ```

Ah, removed, good catch.

> Does that also explain the "Too many distributed transactions for snapshot" error you saw?
Hmm, I don't think so. I saw those errors before I changed the code to 
enlarge the array if needed. Before that, the above code was still 
relevant. Not sure, though. I'm too lazy to check, I think the 
refactoring to enlarge the array on-demand is better in any case. In 
particular, I didn't like the way the QD sent the 'maxCount' to the QE, 
when the only purpose of that was to let the QE to size that 
backend-local array "big enough".

- Heikki

Chibin commented:
If you would like to run the gpcheckcat tests, you can run `behave test/behave/mgmt_utils/gpcheckcat.feature` in the gpMgmt directory.

And if you want to run the unit test, you can o run `python -m unittest2 discover --verbose -s "gppylib" -p "test_unit_gpcheckcat.py"` in the gpMgmt/bin directory.

hlinnaka commented:
Rebased. @ashwinstar, @asimrp, did you have any further comments on this, or is it ready to be pushed?

Does anyone have a test-case at hand for the potential performance issue with the gpcheckcat query?

pivotal-mike commented:
@hlinnaka @asimrp @ashwinstar Let's merge this in for now and leave an issue to revisit the gpcheckcat issue.  There isn't a viable schema at hand to test as requested by @ashwinstar and running tests on a client system is not a viable method.

The work on gp_upgrade will need a schema that should fit the requirements here and once that's in ALPHA we can revisit the gpcheckcat performance issue.

pivotal-mike commented:
For Forms sake:

+1

hlinnaka commented:
Ok, pushed. Thanks for the reviews, everyone! If there was still something to be fixed, please let me know and I'll create a new PR to fix whatever it is.
